mypy/checker.py

```diff
@@ -3788,67 +3788,109 @@ def find_isinstance_check_helper(self, node: Expression) -> Tuple[TypeMap, TypeM
+                    if left_index not in narrowable_operand_indices:
+                        continue
 
-                if is_not:
-                    if_vars, else_vars = else_vars, if_vars
-                return if_vars, else_vars
-            # Check for `x == y` where x is of type Optional[T] and y is of type T
-            # or a type that overlaps with T (or vice versa).
-            elif node.operators == ['==']:
-                first_type = type_map[node.operands[0]]
-                second_type = type_map[node.operands[1]]
-                if is_optional(first_type) != is_optional(second_type):
-                    if is_optional(first_type):
-                        optional_type, comp_type = first_type, second_type
-                        optional_expr = node.operands[0]
+                    item_type = operand_types[left_index]
+                    collection_type = operand_types[right_index]
+
+                    # We only try and narrow away 'None' for now
+                    if not is_optional(item_type):
+                        pass
+
+                    collection_item_type = get_proper_type(builtin_item_type(collection_type))
+                    if collection_item_type is None or is_optional(collection_item_type):
+                        continue
+                    if (isinstance(collection_item_type, Instance)
+                            and collection_item_type.type.fullname == 'builtins.object'):
+                        continue
+                    if is_overlapping_erased_types(item_type, collection_item_type):
+                        if_map, else_map = {operands[left_index]: remove_optional(item_type)}, {}
                     else:
-                        optional_type, comp_type = second_type, first_type
-                        optional_expr = node.operands[1]
-                    if is_overlapping_erased_types(optional_type, comp_type):
-                        return {optional_expr: remove_optional(optional_type)}, {}
-            elif node.operators in [['in'], ['not in']]:
-                expr = node.operands[0]
-                left_type = type_map[expr]
-                right_type = get_proper_type(builtin_item_type(type_map[node.operands[1]]))
-                right_ok = right_type and (not is_optional(right_type) and
-                                           (not isinstance(right_type, Instance) or
-                                            right_type.type.fullname != 'builtins.object'))
-                if (right_type and right_ok and is_optional(left_type) and
-                        literal(expr) == LITERAL_TYPE and not is_literal_none(expr) and
-                        is_overlapping_erased_types(left_type, right_type)):
-                    if node.operators == ['in']:
-                        return {expr: remove_optional(left_type)}, {}
-                    if node.operators == ['not in']:
-                        return {}, {expr: remove_optional(left_type)}
```

Comment:
I can't find where the this whole chunk moved. Do we now support `a in b in c`? If yes, is there a test?

Reply:
This chunk of code starts on line 3853. I guess the diff ended up being a little messy because I renamed the variables ("left_type" -> "item_type" and "right_type" -> "collection_type") and moved the logic around a bit.

But we don't do anything really special for this case: we treat `a in b in c` as if it were `a in b and b in c` and only narrow away Optionals from the LHS item type if it overlaps with whatever's inside the RHS collection.

So in the end, I don't think we've really improved support for this pattern.